alg/sha256.c

```diff
@@ -267,7 +264,19 @@ SHA256_Transform(uint32_t state[static restrict 8],
 	int i;
 
 #ifdef HWACCEL
-	switch(usehw()) {
+
+#if defined(__GNUC__) && defined(__aarch64__)
+	/*
+	 * Optimization for gcc on 64-bit ARM.  This _should_ have no effect,
+	 * because the only place hwaccel is written is inside hwaccel_init(),
+	 * and we have an `assert(hwaccel != HW_UNSET)` immediately after
+	 * calling hwaccel_init() in SHA256_Init().  But evidently that info
+	 * isn't being carried through between functions.
+	 */
+	assert(hwaccel != HW_UNSET);
```

Comment:
There's no way for gcc, or any other compiler, to know that `hwaccel` can't be `HW_UNSET` at this point, because there's no way for them to know that we will always call `SHA256_Init` before calling `SHA256_Transform`.

Just how much of a performance benefit does this provide?

Reply:
It's shown graphically in the image for "arm 64" (in the previous comment), comparing the purple to black bars.

- for a buffer size of 256, `gcc -O2` goes from 560 to 738.  So a 31.7% boost.
- 256 with `gcc -O3` is 563 to 705.  So only a 25% boost.
- buffer size of 1024 is a more modest boost
- buffer size of 10000 is barely noticeable; 1123 to 1137.  1% boost?

It's worth glancing at the graph, just to see how starkly  `gcc` alternates between ~560 to ~720.

Reply:
Ah, got it.  It's not easy (for me at least) to see what's going on when there's three compilers squished together within a single block size on the graph.

Reply:
Yeah, I've tried a bunch of different visualizations, but nothing's jumped out as ideal.

I could try having two levels of bars?  i.e. a very light grey background for each block size, then within that background, I could have medium-light grey bars for each compiler (within which we'd have the colored boxplots)?

Reply:
Ok, I can live with having this assertion here for gcc/arm64.  I'm guessing it hurts performance on other systems and that's why it's a preprocessor conditional?

But the comment needs to be changed to reflect the reality of "we require that SHA256_Init is called before SHA256_Transform, but the compiler has no way of knowing that".

Reply:
It's a minor performance penalty elsewhere (amd64 with `clang -O2` and buffer size of  10000, it's 198 to 197 MB/s; buffer size of 256, it's 178 to 176 MB/s).

More to the point, though, it's a messy hack that's only necessary due to something weird that `gcc` on ARM is doing with the `switch`.  Remember, the `switch` currently contains:
```
case HW_SOFTWARE:
case HW_UNSET:
        break;
```

So why on earth would `assert(hwaccel != HW_UNSET)` have a 31% speed boost?

I've compared the assembly code from both versions, and nothing jumped out at me.  Some code shuffled around, and it's storing values in different registers, so there's a ton of changed lines simply due to those trivial changes.  Also, this gets compiled to multiple locations;
```
$ grep "assert(hwaccel" assembly.difff | wc -l
16
```

... in fact, that gives a clue about the major difference:
```
td@ip-xxx:~/src/libcperciva/tests/sha256$ grep "assert(hwaccel" orig.txt | wc -l
1
td@ip-xxx:~/src/libcperciva/tests/sha256$ grep "assert(hwaccel" new.txt | wc -l
16
```

So evidently the original code is just on the boundary of having certain parts inlined?

If this was amd64, I'd keep on digging.  Or if I saw this behaviour from other compilers.  But at this point, my instinct is that it's not worth investigating further.  There's some quirky behaviour of `gcc` on ARM; we can manage the quirk by adding an `assert`, so good enough for now.

(Of course, if you want me to spend more time on it, I can do that.  gcc has 264 current open bugs which contain "arm"; I could check if any of those are applicable, and if not, try to create a small example of this behaviour.)

Reply:
No need to investigate further.  Just fix the comment.